### PR TITLE
fix base_score problem.

### PR DESCRIPTION
**Pull request by JiaxiangBU**
Here is the reproducible minimal example for testing.

``` r
library(xgboost)
#> Warning: 程辑包'xgboost'是用R版本3.6.1 来建造的
library(tidyverse)
#> Registered S3 methods overwritten by 'ggplot2':
#>   method         from 
#>   [.quosures     rlang
#>   c.quosures     rlang
#>   print.quosures rlang
#> Warning: 程辑包'dplyr'是用R版本3.6.1 来建造的
train_data <- mtcars %>%
  rename(y = am)
dtrain <-
  xgb.DMatrix(data = as.matrix(train_data %>% select(-y))
              , label = train_data$y)
xgb_model <- xgb.train(
  data = dtrain,
  nround = 10,
  seed = 1,
  max_depth = 1,
  objective = "binary:logistic",
  base_score = mean(train_data$y) # fix uncalibration problem
)
pred_from_model <- predict(xgb_model, newdata = dtrain)
library(sqldf)
#> Warning: 程辑包'sqldf'是用R版本3.6.1 来建造的
#> 载入需要的程辑包：gsubfn
#> Warning: 程辑包'gsubfn'是用R版本3.6.1 来建造的
#> 载入需要的程辑包：proto
#> Warning: 程辑包'proto'是用R版本3.6.1 来建造的
#> 载入需要的程辑包：RSQLite
#> Warning: 程辑包'RSQLite'是用R版本3.6.1 来建造的
library(tidypredict)
pred_from_tidypredict <-
  tidypredict_sql(xgb_model, dbplyr::simulate_dbi()) %>%
  paste("select ", ., " from mtcars") %>%
  str_replace("LN", "log") %>% 
  sqldf() %>%
  pull
(pred_from_model - pred_from_tidypredict) %>% abs %>% mean
#> [1] 1.464673e-08
```

<sup>Created on 2019-10-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

**Comment by topepo** [comment]
Please write some test cases for this.

File: R/model-xgboost.R [hunk 1]
```diff
@@ -140,7 +140,7 @@ build_fit_formula_xgb <- function(parsedmodel) {
     assigned <- 1
   } else if (objective %in% c("binary:logistic", "reg:logistic")) {
     assigned <- 1
-    f <- expr(1 - 1 / (1 + exp(!!f)))
+    f <- expr(1 - 1 / (1 + exp(!!f + log(!!base_score/(1-!!base_score)))))
```
[the remark this answers is not in the document]
Comment (topepo) [reply R/model-xgboost.R:143]: I'd like you to add tests to `test_xgboost.R` (including cases with `base_score = 1` or `base_score = 0`). 

I'll need to make a submission to CRAN very soon due to issues with `xgboost`.

[the remark this answers is not in the document]
Comment (topepo) [reply R/model-xgboost.R:143]: You should probably use 

```r
binomial()$linkfun(!!base_score)
```

instead of 

```r
log(!!base_score/(1-!!base_score))))
```

That should take care of the edge cases.

[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/model-xgboost.R:143]: > ```r
> ```r
> log(!!base_score/(1-!!base_score))))
> ```
> 
> 
> That should take care of the edge cases.
> ```

Thanks, I make a change 30c3585, but I think, the base score is hard to equal to 1 or 0 except the target is constant.

[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/model-xgboost.R]: > I'd like you to add tests to `test_xgboost.R` (including cases with `base_score = 1` or `base_score = 0`).
> 
> I'll need to make a submission to CRAN very soon due to issues with `xgboost`.

I write the test block in the `test_xgboost.R` b629775d1936e99f59d0e4821c29f43ae414ec96, and I want to include the edge cases, but I find xgboost already requires basescore from 0 to 1.

```r
> base_score <- 1
> xgb_reglogistic_basescore <-
+   xgboost::xgb.train(
+     params = list(
+       max_depth = 2,
+       silent = 1,
+       objective = "reg:logistic",
+       base_score = base_score
+     ),
+     data = xgb_bin_data,
+     nrounds = 4
+   )
Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj) : 
  [11:52:43] amalgamation/../src/objective/./regression_loss.h:62: Check failed: base_score > 0.0f && base_score < 1.0f: base_score must be in (0,1) for logistic loss
```

[the remark this answers is not in the document]
Comment (topepo) [reply R/model-xgboost.R]: >  xgboost already requires basescore from 0 to 1.

Sounds good. Please wrap that in an `expect_error()`. 

Please add an extra return at the end of the file

image

Also, you might need to update from the master branch; I just fixed some changes related to the latest `xgboost` and your tests might fail without those changes.

[the remark this answers is not in the document]
Comment (JiaxiangBU) [reply R/model-xgboost.R]: > > xgboost already requires basescore from 0 to 1.
> 
> Sounds good. Please wrap that in an `expect_error()`.

I write the four `expect_error` cases for both base scores (0 or 1) and both `reg:logistic` and `binary:logistic` cases. https://github.com/tidymodels/tidypredict/pull/66/commits/4905cc2a83bdf74edd02ed055db8302743a6f83c

> Please add an extra return at the end of the file
> 
> image

Already added. https://github.com/tidymodels/tidypredict/pull/66/commits/24d8d7426c8d74e61c7ef3d5b832bf65332e7ff4

> Also, you might need to update from the master branch; I just fixed some changes related to the latest `xgboost` and your tests might fail without those changes.

Already pulled from the upstream master branch. https://github.com/tidymodels/tidypredict/pull/66/commits/db9defcf5a6d8f1029d99510df06608888a41bc2

**Comment by topepo** [comment]
Thanks for doing this.
